Misleading error when using fully qualified trait syntax without trait

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

Given the following code: <!-- Please provide a link to play.rust-lang.org -->

```rust
trait StringExt {
    fn ext(&self);
}
impl StringExt for String {
    fn ext(&self) {
        <Self as str>::trim(self);
        // str::trim(self); - this works
    }
}
```
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f3ca5eea2c043a3bdf681d6c3e864b67)

The current output is:

```
error[E0576]: cannot find method or associated constant `trim` in `str`
 --> src/lib.rs:6:24
  |
6 |         <Self as str>::trim(self);
  |                        ^^^^ not found in `str`

For more information about this error, try `rustc --explain E0576`.
```

However, `str` is not a trait, therefore the fully-qualified trait syntax should be simply invalid in this case - therefore, the error would better say the real problem ("`str` is not a trait"), probably with the help message suggesting the commented-out line.

Initially found in [this SO question](https://stackoverflow.com/questions/72766027/cannot-find-method-or-associated-constant-join-in-stdthreadjoinhandle).


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"jannden"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

Did you mean the ideal output should like:
```
error[E0576]: cannot find method or associated constant `trim` in `str`
 --> src/lib.rs:6:24
  |
6 |         <Self as str>::trim(self);
  |                        ^^^^ `str` is not a trait

For more information about this error, try `rustc --explain E0576`.
```
?

Current output:

```
error[E0404]: expected trait, found builtin type `str`
 --> src/lib.rs:6:18
  |
6 |         <Self as str>::trim(self);
  |                  ^^^ not a trait
```

--- a/tests/ui/traits/fully-qualified-syntax-cast.rs
+++ b/tests/ui/traits/fully-qualified-syntax-cast.rs
@@ -1,0 +1,15 @@
+// Regression test for #98565: Provide diagnostics when the user uses
+// the built-in type `str` in a cast where a trait is expected.
+
+trait Foo {
+    fn foo(&self);
+}
+
+impl Foo for String {
+    fn foo(&self) {
+        <Self as str>::trim(self);
+        //~^ ERROR expected trait, found builtin type `str`
+    }
+}
+
+fn main() {}

--- a/tests/ui/traits/fully-qualified-syntax-cast.stderr
+++ b/tests/ui/traits/fully-qualified-syntax-cast.stderr
@@ -1,0 +1,9 @@
+error[E0404]: expected trait, found builtin type `str`
+  --> $DIR/fully-qualified-syntax-cast.rs:10:18
+   |
+LL |         <Self as str>::trim(self);
+   |                  ^^^ not a trait
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0404`.

